src/EventStore.Core/TransactionLog/FileNamingStrategy/VersionedPatternFileNamingStrategy.cs

```diff
@@ -44,6 +45,73 @@ public class VersionedPatternFileNamingStrategy : IFileNamingStrategy {
 			return versions;
 		}
 
+		private int GetFileNumberFor(string path) {
+			var fileName = Path.GetFileName(path);
+			if (!_chunkNamePattern.IsMatch(fileName))
+				throw new ArgumentException($"Invalid file name: {fileName}");
+
+			var start = _prefix.Length;
+			var end = fileName.IndexOf('.', _prefix.Length);
+			Debug.Assert(end != -1);
+
+			if (!int.TryParse(fileName[start..end], out var chunkNumber))
+				throw new ArgumentException($"Invalid file name: {fileName}");
+
+			return chunkNumber;
+		}
+
+		private int GetFileVersionFor(string path) {
+			var fileName = Path.GetFileName(path);
+			if (!_chunkNamePattern.IsMatch(fileName))
+				throw new ArgumentException($"Invalid file name: {fileName}");
+
+			var dot = fileName.IndexOf('.', _prefix.Length);
+			Debug.Assert(dot != -1);
+
+			if (!int.TryParse(fileName[(dot+1)..], out var chunkVersion))
+				throw new ArgumentException($"Invalid file name: {fileName}");
+
+			return chunkVersion;
+		}
+
+		public void EnumerateAllFiles(
+			Func<string, int, int, int> getNextFileNumber,
+			Action<string, int, int> onLatestVersionFound,
+			Action<string, int> onOldVersionFound,
+			Action<string, int> onFileMissing) {
+			var allFiles = GetAllPresentFiles();
+			Array.Sort(allFiles, StringComparer.CurrentCultureIgnoreCase);
+
+			int expectedChunkNumber = 0;
+			for (int i = 0; i < allFiles.Length; i++) {
```

Comment:
since we're iterating through the files that exist rather than up to an expected chunk number, theres a problem if fewer files exist than we are expecting, consider:

```
		[Test]
		public void detect_no_database() {
			var config = TFChunkHelper.CreateSizedDbConfig(PathName, 4000, chunkSize: 1000);
			using (var db = new TFChunkDb(config)) {
				Assert.That(() => db.Open(verifyHash: false),
					Throws.Exception.InstanceOf<CorruptDatabaseException>()
						.With.InnerException.InstanceOf<ChunkNotFoundException>());
			}
		}
```

```
  Message: 
  Expected: instance of <EventStore.Core.Exceptions.CorruptDatabaseException> and property InnerException instance of <EventStore.Core.Exceptions.ChunkNotFoundException>
  But was:  no exception thrown
```

Reply:
@timothycoleman great catch! It sounds like passing an expected last file number (derived from the writer checkpoint) should allow us to detect this case.

Reply:
Resolved in eb837db3782c45df817204ecbd8b17142448ff1e